zerver/lib/display_recipient.py

```diff
@@ -122,101 +122,39 @@ def bulk_fetch_user_display_recipients(
-        )
-        for recipient in recipients:
-            if recipient.type == Recipient.PERSONAL:
-                user_ids_to_fetch.add(recipient.type_id)
-            else:
-                user_ids_to_fetch = user_ids_to_fetch.union(huddle_user_ids[recipient.id])
-
-        # Fetch the needed UserProfiles:
-        user_profiles: Dict[int, UserDisplayRecipient] = bulk_get_user_profile_by_id(
-            list(user_ids_to_fetch)
-        )
+    for recipient_id in huddle_recipient_ids:
+        user_ids_to_fetch |= huddle_user_id_dict[recipient_id]
 
-        # Build the return value:
-        result: List[Tuple[int, List[UserDisplayRecipient]]] = []
-        for recipient in recipients:
-            if recipient.type == Recipient.PERSONAL:
-                result.append((recipient.id, [user_profiles[recipient.type_id]]))
-            else:
-                result.append(
-                    (
-                        recipient.id,
-                        [user_profiles[user_id] for user_id in huddle_user_ids[recipient.id]],
-                    )
-                )
-
-        return result
-
-    def personal_and_huddle_cache_transformer(
-        db_object: Tuple[int, List[UserDisplayRecipient]],
-    ) -> List[UserDisplayRecipient]:
-        """
-        Takes an element of the list returned by the query_function, maps it to the final
-        display_recipient list.
-        """
-        user_profile_list = db_object[1]
-        display_recipient = user_profile_list
-
-        return display_recipient
-
-    def personal_and_huddle_id_fetcher(db_object: Tuple[int, List[UserDisplayRecipient]]) -> int:
-        # db_object is a tuple, with recipient_id in the first position
-        return db_object[0]
-
-    # ItemT = Tuple[int, List[UserDisplayRecipient]] (recipient_id, list of corresponding users)
-    # CacheItemT = List[UserDisplayRecipient] (display_recipient list)
-    # ObjKT = int (recipient_id)
-    personal_and_huddle_display_recipients: Dict[
-        int, List[UserDisplayRecipient]
-    ] = transformed_bulk_cached_fetch(
-        cache_key_function=display_recipient_cache_key,
-        query_function=personal_and_huddle_query_function,
-        object_ids=recipient_ids,
-        id_fetcher=personal_and_huddle_id_fetcher,
-        cache_transformer=personal_and_huddle_cache_transformer,
-    )
+    # Fetch the needed user dictionaries.
+    user_display_recipients = bulk_fetch_single_user_display_recipients(list(user_ids_to_fetch))
+
+    result = {}
+
+    for recipient_id, _, user_id in personal_tuples:
+        display_recipients = [user_display_recipients[user_id]]
+        result[recipient_id] = display_recipients
+
+    for recipient_id in huddle_recipient_ids:
+        user_ids = huddle_user_id_dict[recipient_id]
+        display_recipients = [user_display_recipients[user_id] for user_id in user_ids]
+        result[recipient_id] = display_recipients
```

Comment:
Is there a previous or current guarantee that `display_recipients` output is sorted by user ID ascending? I feel like that would be a good property to have for the API, and I kinda suspect the `order_by` in the previous logic attempted to arrange that; I think we could arrange that by just doing `user_ids = sorted(huddle_user_id_dict[recipient_id])` a couple lines above?

Reply:
Fixed.  The `bulk_get_huddle_user_ids` functions returns a list of sets now, so it's entirely possible I broke the order. Technically our clients need to be robust about this, but of course I'd rather be deterministic.